File: November2020/BasicCalculatorII.java
class BasicCalculatorII {
    // TC => O(n), Space O(n)
    public int calculate(String s) {
        Stack<Integer> st = new Stack<>();

        char prevSign = '+';
        int no = 0;
        for(int i=0;i<s.length();i++){
            char c = s.charAt(i);
            if(Character.isDigit(c)){
                no = no*10  + (c-'0');
            }

            if(isOperator(c) || i ==s.length()-1){
                if(prevSign == '+'){
                    st.push(no);
                } else if(prevSign == '-'){
                    st.push(-no);
                } else if(prevSign =='*'){
                    int top = st.pop();

                    st.push(top * no);

                } else if(prevSign =='/'){
                    int top = st.pop();
                    st.push(top/no);
                }
                no =0;
                prevSign =c;
            }
        }

        int res = 0;
        while(!st.empty()){ // 1+2+3+4
            res += st.pop();
        }
        return res;
    }

    private boolean isOperator(char c){
        return c=='+' || c=='-'|| c=='/' || c=='*';
    }
}

File: December2021/C++/Basic_Calculator_II.cpp
class Solution {
public:
    int calculate(string s) {
        stack<int> stack;
        char sign='+';
        int temp=0,num=0,res=0;
        for(int i=0;i<s.length();i++)
        {
            if(isdigit(s[i]))
                temp=temp*10+(s[i]-'0');
            if((!isdigit(s[i]) && !isspace(s[i])) || i==s.length()-1)
            {
                if(sign=='+')
                    stack.push(temp);
                else if(sign=='-')
                    stack.push(-temp);
                else if(sign=='*')
                {
                    num=stack.top()*temp;
                    stack.pop();
                    stack.push(num);
                }
                else
                {
                    num=stack.top()/temp;
                    // cout<<num<<endl;
                    stack.pop();
                    stack.push(num);
                }
                temp=0;
                sign=s[i];
            }
        }
        while(stack.size())
        {
            res+=stack.top(); stack.pop();
        }
        return res;
    }
};